human_aware_rl/imitation/behavior_cloning_tf2.py

```diff
@@ -0,0 +1,358 @@
+                                            as well as the configuration for the rollout (rollout_length)
+
+    Returns
+
+        - reward (int)                  Total sparse reward achieved by AgentPair during rollout
+    """
+    evaluation_params = bc_params['evaluation_params']
+    mdp_params = bc_params['mdp_params']
+
+    # Get reference to state encoding function used by bc agents, with compatible signature
+    base_env = _get_base_env(bc_params)
+    mlp = get_mlp(base_env)
+    def featurize_fn(state):
+        return base_env.mdp.featurize_state(state, mlp)
+
+    # Wrap Keras models in rllib policies
+    agent_0_policy = BehaviorCloningPolicy.from_model(model, bc_params, stochastic=True)
+    agent_1_policy = BehaviorCloningPolicy.from_model(model, bc_params, stochastic=True)
+
+    # Compute the results of the rollout(s)
+    results = evaluate(evaluation_params, mdp_params, agent_0_policy, agent_1_policy, featurize_fn, featurize_fn)
+
+    # Compute the average sparse return obtained in each rollout
+    reward = np.mean(results['ep_returns'])
+    return reward
+
+
+
+################
+# Rllib Policy #
+################
+
+class BehaviorCloningPolicy(RllibPolicy):
+
+    def __init__(self, observation_space, action_space, config):
+        """
+        RLLib compatible constructor for initializing a behavior cloning model
+
+        observation_space (gym.Space|tuple)     Shape of the featurized observations
+        action_space (gym.space|tuple)          Shape of the action space (len(Action.All_ACTIONS),)
+        config (dict)                           Dictionary of relavant bc params
+            - model_dir (str)                   Path to pickled keras.Model used to map observations to action logits
+            - stochastic (bool)                 Whether action should return logit argmax or sample over distribution
+            - bc_model (keras.Model)            Pointer to loaded policy model. Overrides model_dir
+            - bc_params (dict)                  Dictionary of parameters used to train model. Required if "model" is present
+        """
+        super(BehaviorCloningPolicy, self).__init__(observation_space, action_space, config)
+
+        
+
+        if 'bc_model' in config and config['bc_model']:
+            assert 'bc_params' in config, "must specify params in addition to model"
+            assert type(config['bc_model']) == keras.Model, "model must be of type keras.Model"
+            self._graph = None
+            model, bc_params = config['bc_model'], config['bc_params']
+        else:
+            assert 'model_dir' in config, "must specify model directory if model not specified"
+            # Construct a private graph unique to this instance to run forward pass (so it doesn't conflict with rllib graphs)
+            self._graph = tf.Graph()
+            with self._graph.as_default():
+                model, bc_params = load_bc_model(config['model_dir'])
```

Comment:
I agree this is super gross and I initially did exactly what you're suggesting, but it broke rllib (it requires a constructor type to be passed in as a policy, and a static `from_saved_model` function would by of type `bound_method`, so it broke their type-check). Thus, I am required to pass in a policy constructor with the exact signature it currently has in the __init__ function. `from_model_dir` is never actually called anywhere in the training code, since it was rejected by rllib, leading to the overly complex init function

Reply:
I actually opened a github issue with rllib to see if they'll relax the type checking on the custom policy init function, so this might end up getting fixed eventually, but for right now there's really not a much cleaner way to do it

Reply:
I see, ah that's annoying. Oh well, in that case I guess you're right, not much one can do!